### ISSUE TITLE
SubmenuButton submenus close automatically only on first activation for non touchscreen events [Android]

### ISSUE THREAD

**Issue by franciscojma86** [flutter/flutter#154842]
### Steps to reproduce

Create a SubmenuButton with a children menu items
```
  
                      SubmenuButton(
                        leadingIcon: Icon(Symbols.design_services),
                        menuChildren: [
                          MenuItemButton(
                            onPressed: () => print('Pressed'),
                            child: Text('My submenu item'),
                          ),
                        ],
                        child: Text('My submenu'),
                      ),
```

Tap on the Submenu button, and the `My submenu item` will flash open and close immediately. The submenu appears on tap down, and disappears on tap release. This only happens the first time. Any subsequent activations work fine.

### Expected results

Tapping on the submenu will show the children and not disappear.

### Actual results

Tapping on the submenu shows and hides the submenu on tap release.

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: Center(
          child: MenuAnchor(
            builder: (context, controller, child) => TextButton(
              onPressed: () {
                controller.isOpen ? controller.close() : controller.open();
              },
              child: Text('Press here'),
            ),
            menuChildren: [
              SubmenuButton(menuChildren: [
                MenuItemButton(
                  onPressed: () {},
                  child: Text('Menu item'),
                ),
              ], child: Text('Submenu')),
            ],
          ),
        ),
      ),
    );
  }
}
```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[Paste your output here]
```

</details>

**Comment by darshankawar** [comment]
Thanks for the report. I was able to replicate this on Android only, as shown below:


https://github.com/user-attachments/assets/9e44eb51-8d4f-49d9-aa38-abd840f95f8e

The same feature / behavior works properly on iOS.

```
Stable : 3.24.2
Master: 3.25.0-1.0.pre.319
```

**Comment by sethladd** [comment]
Set to P1 after conversation with customer, we can dig in more to understand timeline/urgency with customer.

**Comment by franciscojma86** [comment]
Internal issue: b/362646291

**Comment by jason-simmons** [comment]
This started with https://github.com/flutter/flutter/commit/1f4c6ebc972a73a60b4bdeb331c8b5137bd95a3d

**Comment by TahaTesser** [comment]
Taking a look into the issue.

**Comment by TahaTesser** [comment]
I've filed a separate issue for the `MouseRegion` callbacks issue on Android.  https://github.com/flutter/flutter/issues/155021

**Comment by TahaTesser** [comment]
cc: @reidbaker

**Comment by jason-simmons** [comment]
This issue happens on Android because `PointerDataPacketConverter::ConvertPointerData` thinks that it needs to generate a hover event the second time that the user taps the submenu.

(see https://github.com/flutter/engine/blob/main/lib/ui/window/pointer_data_packet_converter.cc#L155)

On the first tap the Android embedding sends a sequence of `PointerChange.add`, `PointerChange.down`, `PointerChange.move`, and `PointerChange.up` events.

However, there is no `PointerChange.remove` event when the user finishes the tap.  So the `PointerDataPacketConverter` still thinks that the pointer is active.
When the `PointerDataPacketConverter` receives the `PointerData::Change::kDown` at the start of the second tap, `LocationNeedsUpdate` thinks that the preexisting pointer has moved and tries to represent this by adding a hover event.  This produces the `MouseRegion` callback that confuses the `SubmenuButton`.

On iOS each tap is completed with a `PointerChange.remove` event that clears the `PointerDataPacketConverter` state.

The `SubmenuButton` works as expected on Android if I patch `AndroidTouchProcessor.onTouchEvent` to add a `PointerChange.remove` event after the `PointerChange.up`.

**Comment by TahaTesser** [comment]
@jason-simmons thank you! If this is fixed on the engine side. I'll be able a regression test for this menu anchor scenario on the framework side. 

I'm wondering if we will wait for engine or we could somehow land my fix right now to unblock bug?

**Comment by bleroux** [comment]
Based on Jason's comment, I updated Taha's test to make it fails without the fix and pass with it:

```dart

  // Regression test for https://github.com/flutter/flutter/issues/154842
  testWidgets('SubmenuButton mouse onHover is not triggered on Android', (WidgetTester tester) async {
    await tester.pumpWidget(MaterialApp(
      home: Material(
        child: MenuAnchor(
          builder: (BuildContext context, MenuController controller, Widget? child) => TextButton(
            onPressed: () {
              controller.isOpen ? controller.close() : controller.open();
            },
            child: const Text('Tap me'),
          ),
          menuChildren: <Widget>[
            SubmenuButton(
              menuChildren: <Widget>[
                MenuItemButton(
                  onPressed: () {},
                  child: const Text('Menu item'),
                ),
              ],
              child: const Text('Submenu'),
            ),
          ],
        ),
      ),
    ));

    // Touch the button to open the menu.
    await tester.tap(find.text('Tap me'));
    await tester.pump();

    // Check that the submenu is open.
    expect(find.text('Submenu'), findsOneWidget);

    // Tap the submenu button.
    final Offset subMenuPhysicalCenter = tester.getCenter(find.text('Submenu')) * tester.view.devicePixelRatio;
    ui.PointerDataPacket packet = ui.PointerDataPacket(
      data: <ui.PointerData>[
        ui.PointerData(change: ui.PointerChange.hover, physicalX: subMenuPhysicalCenter.dx, physicalY: subMenuPhysicalCenter.dy),
        ui.PointerData(change: ui.PointerChange.down, physicalX: subMenuPhysicalCenter.dx, physicalY: subMenuPhysicalCenter.dy),
        ui.PointerData(change: ui.PointerChange.up, physicalX: subMenuPhysicalCenter.dx, physicalY: subMenuPhysicalCenter.dy),
      ],
    );
    GestureBinding.instance.platformDispatcher.onPointerDataPacket?.call(packet);
    await tester.pump();

    // Check that the menu item is open.
    expect(find.text('Menu item'), findsOneWidget);
  }, variant: TargetPlatformVariant.only(TargetPlatform.android));

```

**Comment by gmackall** [comment]
Working on a PR to add a `PointerChange.remove` event after the `UP` - just trying to determine how to handle the case of multiple pointers at the moment

**Comment by TahaTesser** [comment]
I updated the regression test for the frameowkr in https://github.com/flutter/flutter/pull/155023. 

@gmackall when your engine fix lands I can remove my framework fix and just land the updated framework regression test. (Thanks t @bleroux!)

**Comment by dkwingsmt** [comment]
https://github.com/flutter/engine/pull/55157 should fix this problem for touch screens, but doesn't for styli. If the user puts away the stylus and then use the stylus to press a button, then the button receives a HOVER event and then a DOWN event, which opens and closes the menu back to back. 

However, we should also distinguish it from another case, where the user hovers the stylus onto a button, then presses, which I imagine _should_ opens and closes the menu. However, I don't know if my assumption is correct. Does anyone own an Android tablet to confirm?

**Comment by gmackall** [comment]
I am not currently working on this, after fixing the issue for touchscreen events. Unassigning and lowering the priority to p2 to reflect
